### PR TITLE
Refactor CodeBlockSyntax

### PR DESCRIPTION
**Pull request by chenzhiguang**
To have 100% strict.

### CODE REVIEW

File: lib/src/util.dart [hunk 1]
```diff
@@ -164,3 +164,55 @@ String escapePunctuation(String input) {
...
+  bool get isBlank => trim().isEmpty;
+}
+
+class IndentedText {
+  final String text;
+
+  /// How many spaces of a tab that remains after part of it has been consumed.
+  ///
+  /// `null` means it did not hit a `tab`.
+  final int? tabRemaining;
```
Comment (chenzhiguang) [inline lib/src/util.dart:215]: I will need this `tabRemaining` in another syntax.

And here is an explanation about this `tabRemaining`:

For example this an indented code block which starts with 3 spaces and a tab.

`···\t`var a = 1

A table represents 4 spaces,  one of them is consumed by the indented code block structure, so there are 3 spaces left, I call these 3 spaces `tabRemaining`

File: lib/src/block_syntaxes/code_block_syntax.dart [hunk 2]
```diff
@@ -59,4 +55,23 @@ class CodeBlockSyntax extends BlockSyntax {
 
     return Element('pre', [Element.text('code', content)]);
   }
+
+  bool _shouldEnd(BlockParser parser) {
+    var i = 1;
+    while (true) {
+      final nextLine = parser.peek(i);
+      if (nextLine == null) {
```
Comment (srawlins) [inline lib/src/block_syntaxes/code_block_syntax.dart:63]: Can you add a comment here on what this means, (I assume it means EOF, but I'm not sure).

File: lib/src/util.dart [hunk 3]
```diff
@@ -131,3 +131,55 @@ String escapePunctuation(String input) {
 
   return buffer.toString();
 }
+
+extension StringExtensions on String {
+  /// Removes leading whitespace by the length of [length].
```
Comment (srawlins) [inline lib/src/util.dart:136]: Maybe "Removes up to [length] characters of leading whitespace."

File: lib/src/util.dart [hunk 4]
```diff
@@ -131,3 +131,55 @@ String escapePunctuation(String input) {
...
+  // The way of handling tabs: https://spec.commonmark.org/0.30/#tabs
+  IndentedText indent([int length = 4]) {
```
Comment (srawlins) [inline lib/src/util.dart:138]: I think this should be called _dedent_. "indent" would _add_ whitespace.

File: lib/src/util.dart [hunk 5]
```diff
@@ -131,3 +131,55 @@ String escapePunctuation(String input) {
...
+      }
+    }
+    return IndentedText(substring(start), tabRemaining);
+  }
+
+  /// Whether this string contains only whitespaces.
+  bool get isBlank => trim().isEmpty;
+}
+
+class IndentedText {
```
Comment (srawlins) [inline lib/src/util.dart:176]: Please add a doc comment for this new class.

File: lib/src/util.dart [hunk 6]
```diff
@@ -131,3 +131,55 @@ String escapePunctuation(String input) {
...
+  final String text;
+
+  /// How many spaces of a tab that remains after part of it has been consumed.
+  ///
+  /// `null` means it did not hit a `tab`.
```
Comment (srawlins) [inline lib/src/util.dart:181]: How about "`null` means we did not read a 'tab'`".

**Review by srawlins — CHANGES_REQUESTED** [review]

File: lib/src/util.dart (hunk 4 above)
Comment (chenzhiguang) [reply lib/src/util.dart]: Fixed. thanks!

File: lib/src/util.dart (hunk 5 above)
Comment (chenzhiguang) [reply lib/src/util.dart]: Fixed

File: lib/src/util.dart (hunk 6 above)
Comment (chenzhiguang) [reply lib/src/util.dart]: Fixed!

File: lib/src/util.dart (hunk 3 above)
Comment (chenzhiguang) [reply lib/src/util.dart]: Fixed, thanks!

**Review by srawlins — APPROVED** [review]
Thanks

**Pull request merged**